src/core/tUnification.ml

```diff
@@ -889,48 +910,56 @@ and unify_to uctx a b ab tl =
 	if not (unifies_to uctx a b ab tl) then error [cannot_unify a b]
 
 and unifies_abstracts uctx a b a1 tl1 a2 tl2 =
-	let uctx_no_transitive_casts = {uctx with allow_transitive_cast = false} in
-	(unifies_to uctx_no_transitive_casts a b a1 tl1) || (unifies_from uctx_no_transitive_casts a b a2 tl2)
-	|| (((Meta.has Meta.CoreType a1.a_meta) || (Meta.has Meta.CoreType a2.a_meta))
-		&& ((unifies_to uctx a b a1 tl1) || (unifies_from uctx a b a2 tl2)))
+	unifies_with_abstract uctx a b (fun uctx ->
+		List.exists (unifies_to_direct uctx a b a1 tl1) a1.a_to ||
+		List.exists (unifies_from_direct uctx a b a2 tl2) a2.a_from
+	)
 
 and unifies_from uctx a b ab tl =
-	List.exists (unifies_from_direct uctx a b ab tl) ab.a_from
+	unifies_with_abstract uctx a b (fun uctx ->
+		List.exists (unifies_from_direct uctx a b ab tl) ab.a_from
+	)
 
 and unifies_to uctx a b ab tl =
-	List.exists (unifies_to_direct uctx a b ab tl) ab.a_to
+	unifies_with_abstract uctx a b (fun uctx ->
+		List.exists (unifies_to_direct uctx a b ab tl) ab.a_to
+	)
 
 and unifies_from_direct uctx a b ab tl t =
-	rec_stack_abstract_unifies a b (fun() ->
+	does_func_unify (fun() ->
```

Comment:
Doesn't this lose the stack check?

Reply:
If you use those functions from outside `unify` (for field cast search) then you will skip adding first `(a,b)` pair (there is nothing to check at this point), otherwise it is the same. But by moving checking up many redundant equality checks are removed.